trackers/core/reid/model.py

```diff
@@ -0,0 +1,360 @@
+        self,
+        train_loader: DataLoader,
+        epochs: int,
+        validation_loader: Optional[DataLoader] = None,
+        projection_dimension: Optional[int] = None,
+        freeze_backbone: bool = False,
+        optimizer_class: str = "torch.optim.Adam",
+        learning_rate: float = 5e-5,
+        optimizer_kwargs: dict[str, Any] = {},
+        checkpoint_interval: Optional[int] = None,
+        checkpoint_dir: str = "checkpoints",
+        log_to_tensorboard: bool = False,
+        log_to_wandb: bool = False,
+    ) -> None:
+        """
+        Train/fine-tune the ReID model.
+
+        Args:
+            train_loader (DataLoader): The training data loader.
+            epochs (int): The number of epochs to train the model.
+            validation_loader (Optional[DataLoader]): The validation data loader.
+            projection_dimension (Optional[int]): The dimension of the projection layer.
+            freeze_backbone (bool): Whether to freeze the backbone of the model. The
+                backbone is only frozen if `projection_dimension` is specified.
+            optimizer_class (str): The optimizer class to use.
+            learning_rate (float): The learning rate to use for the optimizer.
+            optimizer_kwargs (dict[str, Any]): The optimizer kwargs to use.
+            checkpoint_interval (Optional[int]): The interval to save checkpoints.
+            checkpoint_dir (str): The directory to save checkpoints.
+            log_to_tensorboard (bool): Whether to log to tensorboard.
+            log_to_wandb (bool): Whether to log to wandb. If `checkpoint_interval` is
+                specified, the model will be logged to wandb as well.
+                Project and entity name should be set using the environment variables
+                `WANDB_PROJECT` and `WANDB_ENTITY`. For more details, refer to
+                [wandb environment variables](https://docs.wandb.ai/guides/track/environment-variables).
+        """
+        os.makedirs(checkpoint_dir, exist_ok=True)
+
+        self.add_projection_layer(projection_dimension, freeze_backbone)
+
+        # Initialize optimizer and criterion
+        self.optimizer = eval(optimizer_class)(
+            self.backbone_model.parameters(), lr=learning_rate, **optimizer_kwargs
+        )
+        self.criterion = nn.TripletMarginLoss(margin=1.0)
+
+        config = {
+            "epochs": epochs,
+            "learning_rate": learning_rate,
+            "optimizer_class": optimizer_class,
+            "optimizer_kwargs": optimizer_kwargs,
+            "projection_dimension": projection_dimension,
+            "freeze_backbone": freeze_backbone,
+            "model_metadata": self.model_metadata,
+        }
+
+        # Initialize callbacks
+        callbacks: list[BaseCallback] = []
+        if log_to_tensorboard:
+            callbacks.append(TensorboardCallback())
+        if log_to_wandb:
+            callbacks.append(WandbCallback(config=config))
```

Comment:
We should make the imports for `TensorboardCallback` and `WandbCallback` optional, so that if `tensorboard` or `wandb` aren't installed, the user gets a clear warning suggesting they install the appropriate extra dependencies.